### PR TITLE
PUBDEV-8616: make language rules generation on demand + introduce EnumLimited encoding

### PR DESCRIPTION
**Pull request by koniecsveta**
this change will ensure that the language rule representations are created only on demand for significant rules, and not during the training for all the rules.

https://h2oai.atlassian.net/browse/PUBDEV-8616

### CODE REVIEW

**Comment by valenad1** [comment]
Can we have test for python API?

File: h2o-algos/src/main/java/hex/rulefit/Condition.java [hunk 1]
```diff
@@ -62,12 +61,18 @@ String constructLanguageCondition() {
 
     @Override
     public boolean equals(Object obj) {
```
Comment (valenad1) [inline h2o-algos/src/main/java/hex/rulefit/Condition.java:63]: Can we have a separate test for this?

File: h2o-algos/src/main/java/hex/rulefit/Rule.java [hunk 2]
```diff
@@ -54,19 +52,18 @@ public void map(Chunk[] cs, byte[] out) {
...
 
     @Override
     public boolean equals(Object obj) {
-        return this.hashCode() == obj.hashCode();
+        if (!(obj instanceof Rule))
+            return false;
+
+        Rule rule = (Rule) obj;
+        for (int i = 0; i < conditions.length; i++) {
+            if (!(Arrays.stream(rule.conditions).anyMatch(conditions[i]::equals)))
```
Comment (valenad1) [inline h2o-algos/src/main/java/hex/rulefit/Rule.java:63]: Can we create stream just once and use it multiple time in loop?

**Review by valenad1 — COMMENTED** [review]
Looks good! Thank you! I miss some tests and I would ask 
#h2o-3-api-updates for the `rule_importance`

File: h2o-algos/src/main/java/hex/rulefit/Condition.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Condition.java:63]: it's tested in RuleFitUtilsTest.java#testConditionAndRule()

**Comment by koniecsveta** [comment]
> Looks good! Thank you! I miss some tests and I would ask
> #h2o-3-api-updates for the `rule_importance`

rule_importance() is already present in py client, it's not a new method. The change is, what is being returned. After this change, the method will be present also in R, to be consistent with py.

File: h2o-algos/src/main/java/hex/rulefit/Rule.java [hunk 3]
```diff
@@ -54,19 +52,18 @@ public void map(Chunk[] cs, byte[] out) {
...
-    public int hashCode() {
-        int hashCode = 0;
-        for (int i = 0; i < conditions.length; i++) {
-            hashCode += conditions[i].hashCode();
-        }
-        return hashCode;
-    }
 
     @Override
     public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Rule.java:57]: extended RuleFitUtilsTest.java#testConditionAndRule()

File: h2o-algos/src/main/java/hex/rulefit/Rule.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Rule.java:63]: No, stream can be consumed only once in Java. But I did some adjustments to make it nicer.

**Comment by koniecsveta** [comment]
> Can we have test for python API?

We already have some tests for example in pyunit_rulefit_titanic.py / runit_rulefit_titanic.R

**Review by valenad1 — APPROVED** [review]
All right, thank you!

**Review by tomasfryda — APPROVED** [review]
LGTM, thank you @koniecsveta !

**Comment by koniecsveta** [comment]
@valenad1 , @tomasfryda can you please re-review? I did multiple changes so that everything is working.. Thanks!

**Review by tomasfryda — APPROVED** [review]
Thank you @koniecsveta !

**Review by valenad1 — APPROVED** [review]
Still good for me 👍

**Review by michalkurka — APPROVED** [review]

**Pull request merged**